addons/payment/models/payment_acquirer.py

```diff
@@ -230,8 +255,18 @@ def create(self, values_list):
         return acquirers
 
     def write(self, values):
-        result = super().write(values)
         self._check_required_if_provider()
+
+        if any(field in ['state', 'allow_tokenization'] for field in values):
+            for acquirer in self:
+                if acquirer.state != values.get('state', acquirer.state) or (
+                    acquirer.allow_tokenization and not values.get('allow_tokenization', True)
+                ):
+                    self.env['payment.token'].search([
+                        ('acquirer_id', 'in', self.ids),
+                    ]).write({'active':False})
+
+        result = super().write(values)
```

Comment:
Why can't you apply your logic post-write ?
(FYI, `_check_required_if_provider` must be called after the super write call because it needs to verify the acquirer values after the update, not before).

Reply:
```python
modified_acquirers = self.env['payment.acquirer']
if not values.get('allow_tokenization', True):
    modified_acquirers |= self.filtered('allow_tokenization')
if values.get('state'):
    modified_acquirers |= self.filtered(lambda acq: acq.state != values.get('state'))
   
res = super().write(values)
if modified_acquirers:
    self.env['payment.token'].search(
        ('acquirer_id', 'in', modified_acquirers.ids),
        ('active', '=', True),
    ]).write({'active':False})

return res
```
?
* post-write
* all tokens are archived at once :rocket:

Reply:
Okay, now I see the benefit. I didn't realize that tokens were not archived at once. I'll go with your solution